### PR TITLE
Write archives from packaging

### PR DESCRIPTION
**Pull request by jasonrclark**
This PR revises `furoshiki` to write out archive files as its output (`tar.gz` for Mac and Linux, `zip` for Windows) rather than loose files in directories. In addition to being a nice-to-have for users, this addresses the problem from https://github.com/shoes/shoes4/issues/1448 because the tar retains executable permissions even when built on Windows.

Other highlights:
* Changed the shared contexts to be actual RSpec style instead of using instance variables
* Deleted a bunch of zip related stuff only lightly used for testing
* Removed unused helpers

TODO:
* [x] TESTING! Plenty of cross-platform testing to do before I'll be ready to merge

### CODE REVIEW

File: lib/furoshiki/base_app.rb [hunk 1]
```diff
@@ -182,13 +199,26 @@ def inject_jar
...
 
-    def after_moved
-    end
-
-    def move_to_package_dir(path)
-      dest = package_dir.join(app_name)
-      dest.rmtree if dest.exist?
-      mv path.to_s, dest
+    def create_archive(source_path)
+      dest = package_dir.join(archive_name)
```
Comment (PragTob) [inline lib/furoshiki/base_app.rb]: I think `destination` helps a lot here :)

File: lib/furoshiki/base_app.rb [hunk 2]
```diff
@@ -182,13 +199,26 @@ def inject_jar
...
+      File.open(dest, "wb") do |file|
+        Zlib::GzipWriter.wrap(file) do |gz|
+          Gem::Package::TarWriter.new(gz) do |tar|
+            tmp_files.each do |source_item|
+              dest_item = source_item.sub(source_path.to_s, app_name)
```
Comment (PragTob) [inline lib/furoshiki/base_app.rb]: really a fan of the longer names :)

Comment (PragTob) [inline lib/furoshiki/base_app.rb]: If I see this correctly we only need `tar` and `source_item` within the block. Imo we could get all the File, Zlib and Tar  stuff into its own method (maybe also the tmp_files iteration) as the whole wrapping of the file doesn't matter much for what happens inside this, right? Then the method would be much shorter and not so indented :)

File: lib/furoshiki/zip/recursive.rb [hunk 3]
```diff
@@ -1,58 +0,0 @@
...
-            subdir = disk_path.children(false)
-            write_entries(subdir, disk_prefix, zip_path, io)
-          else
-            io.get_output_stream(zip_path) { |f| f.puts(File.open(disk_path, "rb").read())}
-          end
-        end
-      end
-    end
-  end
-end
```
Comment (PragTob) [inline lib/furoshiki/zip/recursive.rb]: :+1:

**Comment by jasonrclark** [comment]
Great comments @PragTob, and thanks for taking a look! I've decomposed and renamed through the `create_archive` method quite a bit and I like how it ended up much better. Review FTW 💖

**Pull request merged**